### DISCUSSION TITLE
Formatting of multiline strings

### DISCUSSION THREAD

**Discussion by 0x6675636b796f75676974687562** [cqfn/diKTat#1372] [Ideas]
# Abstract
In light of the recent fixes (#1364)) for the way we format multiline strings, it makes sense to agree on how exactly we format string literals which are not effective call arguments.

There're two aspects to consider:

1. The indentation of the closing `"""`.
2. The indentation of single `LITERAL_STRING_TEMPLATE_ENTRY` nodes (at least, in the `.trimMargin()` case).

# Ideas

Variants:

1. The closing `"""` as well as each line are aligned using a fractional indent size (not a multiple of 4, we can introduce a separate flag similar to `alignedParameters`):
    ```kotlin
    fun f() {
        val xyz = """
                  | First
                  | Second
                  | Third
                  """.trimMargin()
    }
    ```
    This approach may be considered similar to the way block/Javadoc/KDoc/Doxygen comments are formatted, and there're variations, too:
    ```kotlin
    /*
     * Either this...
     */

    /*
    * ... or this...
    */

    /**
      * ... or even this (Javadoc/KDoc/Doxygen only).
      */
    ```
2. The closing `"""` is indented, using a _single_ or a _continuation_ indent, w. r. t. the current indentation level, and string fragments which start with `|`, use the same indentation level (we'd need a new `extendedIndentInMultilineStrings` flag):
    ```kotlin
    companion object Foo : Bar {
        private val xyz = """
            | First
            | Second
            | Third
            """.trimMargin()
    }
    ```
    Same as above, using a _continuation_ indent (`extendedIndentInMultilineStrings` is **on**):
    ```kotlin
    companion object Foo : Bar {
        private val xyz = """
                | First
                | Second
                | Third
                """.trimMargin()
    }
    ```
3. Neither the closing `"""` nor any fragments starting with `|` are indented at all:
    ```kotlin
    companion object {
        private val xyz = """
        | First
        | Second
        | Third
        """.trimMargin()
    }
    ```

All the above examples may feature `|`-fragments being additionally indented w. r. t. the closing `"""`, optionally controlled by a separate flag, e. g.:
```kotlin
    """
        | First
        | Second
        | Third
    """.trimMargin()
```

Whenever a string literal is followed by a `.trimIndent()` call or a `.trimXyz()` call or **no call at all** (we must handle this case, too), `LITERAL_STRING_TEMPLATE_ENTRY` nodes should **retain their original indentation**, only the closing `"""` should be indented.

# Existing style guides

Kotlin [style guide](https://kotlinlang.org/docs/coding-conventions.html#strings) says nothing about indentation (nor does IDEA) but provides these examples:

```kotlin
println("""
    Not
    trimmed
    text
    """
       )

println("""
    Trimmed
    text
    """.trimIndent()
       )

println()

// A.S.: LITERAL_STRING_TEMPLATE_ENTRY nodes indented using 10 spaces despite 4 spaces are used for indentation:
val a = """Trimmed to margin text:
          |if(a > 1) {
          |    return a
          |}""".trimMargin()

println(a)
```

Additionally, the Java-to-Kotlin [migration guide](https://kotlinlang.org/docs/java-to-kotlin-idioms-strings.html#use-multiline-strings) provides two extra examples:

```kotlin
// Kotlin   
val result = """
        Kotlin
        Java 
    """.trimIndent()
println(result)
```

(the closing `"""` indented once) and 

```kotlin
// Kotlin
fun main() {
    val result = """
       #  Kotlin
       #  Java
   """.trimMargin("#")
    println(result)
}
```

(the closing `"""` indented using 3 spaces which is probably a mistake).

@akuleshov7, @Arrgentum, WDYT?

**Comment by orchestr7** [comment]
I love this one:
```
val a = // <-- new line?
    """
        | First
        | Second
        | Third
    """.trimMargin()
```

May be we will force to move it to a new line ALWAYS?


This one is also fine:
```
val result = """
        Kotlin
        Java 
    """.trimIndent()
```

But I prefer the first one

**Reply by 0x6675636b796f75676974687562** [reply]
Good point, but how about single-line raw strings?

```kotlin
val a = """|Mary had a little lamb""".trimMargin()
```

**Reply by orchestr7** [reply]
Change it to a single line text instead of multiline with diktat :D
Can be a separate inspection 

```
val a = "|Mary had a little lamb".trimMargin()
```

**Reply by petertrr** [reply]
Single-quoted and triple-quoted strings have different escaping rules, so it may be quite complicated to implement (and in some cases users would use triple-quoted strings just because of this):
```kotlin
val a = """|Mary had a "little" lamb""".trimMargin()  // <-- we'll have to find quotes inside the triple-quoted string and escape them for correct conversion
```

**Reply by orchestr7** [reply]
yeah, then we can simply leave it as is
